Pawelka/linux tests

moozzyk opened the pull request.

moozzyk commented:
/cc @BrennanConroy @halter73

BrennanConroy commented on include/signalrclient/hub_proxy.h:
```diff
@@ -74,4 +61,17 @@ namespace signalr
     };
+
+    template<>
+    inline pplx::task<void> hub_proxy::invoke<void>(const utility::string_t& method_name, const on_progress_handler& on_progress)
```
any particular reason for this change?

moozzyk replied:
Yes. It won't compile with gcc. From what I remember when I hit this it turned out that the compiler in VS allows both but only the definition outside the class is per standard and the definition inside class is an extension

BrennanConroy commented:
How do you build on Linux? I tried using CMake but hitting some errors

moozzyk commented:
@BrennanConroy - I will be adding building instructions to README but in general you do this:
1) build casablanca as per https://github.com/Microsoft/cpprestsdk/wiki/How-to-build-for-Linux (note I built version 2.6.0 but 2.7.0 is now available which I have not tried yet)
2) create a subfolder in the SignalR root repo (e.g. build)
3) run  `CXX=g++-4.8 cmake .. -DCMAKE_BUILD_TYPE=Release -DCPPREST_INCLUDE_DIR=~/source/casablanca/Release/include/ -DCPPREST_LIB_DIR=~/source/casablanca/Release/build.release/Binaries` (adjust values of CPPREST_INCLUDE_DIR and CPPREST_LIB_DIR accordingly)
4) make
This will build the lib and the tests. The binaries will be in the bin subfolder.

There are some issues with the tests: 
 - there are 3 tests that are flakey and fail quite often (especially when compiled in Debug)
 - more often than not (again especially in the debug) - all tests appear to have finished but the process does not exit. Looks like gtest is waiting for some threads to finish
 - My Linux box is a quite old laptop so there might be more (or fewer?) issues on faster boxes 

Thanks for looking at this.

halter73 commented:
@moozzyk I haven't looked at this super carefully, but it seems good to me. Do you want someone else to try running this? I have an Arch Linux VM which might show us something new.

moozzyk commented:
@halter73 - Thanks for taking a look. It would be great to try it in a different setup I tried so far (keeping in mind that Linux support is experimental at this point - see my comment about tests).

BrennanConroy commented:
I'm hitting an error with the make_unique on line 27 of connection_implementation.cpp

Sent from my HTC

----- Reply message -----
From: "Pawel Kadluczka" <notifications@github.com>
To: "aspnet/SignalR-Client-Cpp" <SignalR-Client-Cpp@noreply.github.com>
Cc: "Brennan Conroy" <brecon@microsoft.com>
Subject: [SignalR-Client-Cpp] Pawelka/linux tests (#132)
Date: Fri, Dec 11, 2015 7:34 PM


@halter73<https://github.com/halter73> - Thanks for taking a look. It would be great to try it in a different setup I tried so far (keeping in mind that Linux support is experimental at this point - see my comment about tests).

-
Reply to this email directly or view it on GitHub<https://github.com/aspnet/SignalR-Client-Cpp/pull/132#issuecomment-164105765>.

moozzyk commented:
Weird - I did not see it. I will try to `git clean` my tree and rebuild. I will also check if the include is not missing since I had to do `make_unique` on my own.

moozzyk commented:
@BrennanConroy - I fixed the issue with make_unique and moved to cpprestsdk 2.7.0.
I also wanted to try the client on CentOs but could not build casablanca - it could not find the correct version of (suprise) OpenSSL

BrennanConroy commented:
Looks good, seeing these 2 errors
```
[  FAILED  ] connection_impl_start.connection_state_is_connecting_when_connection_is_being_started
[  FAILED  ] connection_impl_reconnect.reconnect_cancelled_if_connection_dropped_during_start_and_start_failed
```

moozzyk commented:
Thanks @BrennanConroy - I filed a bug for this: https://github.com/aspnet/SignalR-Client-Cpp/issues/136
